src/github_runner_image_builder/builder.py

```diff
@@ -0,0 +1,553 @@
+        )
+        subprocess.run(  # nosec: B603
+            ["/usr/sbin/usermod", "-aG", MICROK8S_GROUP, UBUNTU_USER], check=True, timeout=30
+        )
+        subprocess.run(  # nosec: B603
+            ["/usr/sbin/usermod", "-aG", LXD_GROUP, UBUNTU_USER], check=True, timeout=30
+        )
+        subprocess.run(  # nosec: B603
+            ["/usr/bin/chmod", "777", "/usr/local/bin"], check=True, timeout=30
+        )
+    except subprocess.SubprocessError as exc:
+        raise SystemUserConfigurationError from exc
+
+
+def _install_external_packages() -> None:
+    """Install packages outside of apt.
+
+    Installs yarn.
+
+    Raises:
+        ExternalPackageInstallError: If there was an error installing external package.
+    """
+    try:
+        # 2024/04/26 There's a potential security risk here, npm is subject to toolchain attacks.
+        subprocess.run(
+            ["/usr/bin/npm", "install", "--global", "yarn"], check=True, timeout=60 * 5
+        )  # nosec: B603
+        subprocess.run(
+            ["/usr/bin/npm", "cache", "clean", "--force"], check=True, timeout=60
+        )  # nosec: B603
+    except subprocess.SubprocessError as exc:
+        raise ExternalPackageInstallError from exc
+
+
+@retry(tries=5, delay=5, max_delay=60, backoff=2, local_logger=logger)
+def _compress_image(image: Path) -> None:
+    """Compress the cloud image.
+
+    Args:
+        image: The image to compress.
+
+    Raises:
+        ImageCompressError: If there was something wrong compressing the image.
+    """
+    try:
+        subprocess.run(  # nosec: B603
+            ["/usr/bin/virt-sparsify", "--compress", str(image), str(IMAGE_OUTPUT_PATH)],
+            check=True,
+            timeout=60 * 10,
+        )
+    except subprocess.CalledProcessError as exc:
+        raise ImageCompressError from exc
+
+
+@dataclasses.dataclass
+class BuildImageConfig:
+    """Configuration for building the image.
+
+    Attributes:
+        arch: The CPU architecture to build the image for.
+        base_image: The ubuntu image to use as build base.
+    """
+
+    arch: Arch
+    base_image: BaseImage
+
+
+def build_image(config: BuildImageConfig) -> None:
+    """Build and save the image locally.
+
+    Args:
+        config: The configuration values to build the image with.
+
+    Raises:
+        BuildImageError: If there was an error building the image.
+    """
+    logger.info("Clean build state.")
+    with redirect_stdout(sys.stderr):
+        try:
+            _clean_build_state()
+            logger.info("Downloading cloud image.")
+            cloud_image_path = _download_cloud_image(
+                arch=config.arch, base_image=config.base_image
+            )
+            logger.info("Resizing cloud image.")
+            _resize_cloud_img(cloud_image_path=cloud_image_path)
+            logger.info("Mounting network block device.")
+            _mount_image_to_network_block_device(cloud_image_path=cloud_image_path)
```

Comment:
rename `mount_image_to_nbd`, `cloud_image` rename to `base_image` everywhere.
arg cloud_image_path -> image-path